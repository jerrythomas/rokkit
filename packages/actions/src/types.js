/**
 * @typedef SvelteActionReturn
 * @property {() => void} destroy
 * @property {() => void} [update]
 */

/**
 * @typedef FillableData
 * @property {string} value
 * @property {integer} actualIndex
 * @property {integer} expectedIndex
 */

/**
 * @typedef FillOptions
 * @property {Array<FillableData>} options available options to fill
 * @property {integer} current       index of option to be filled
 * @property {boolean} check         validate filled values
 */

/**
 * A part of the path to node in hierarchy
 *
 * @typedef PathFragment
 * @property {integer}                             index  - Index to item in array
 * @property {Array<*>}                            items  - Array of items
 * @property {import('@rokkit/core').FieldMapping} fields - Field mapping for the data
 */

/**
 * Options for the Navigable action
 * @typedef NavigableOptions
 * @property {boolean} horizontal - Navigate horizontally
 * @property {boolean} nested     - Navigate nested items
 * @property {boolean} enabled    - Enable navigation
 */

/**
 * @typedef NavigatorOptions
 * @property {Array<*>}     items           - An array containing the data set to navigate
 * @property {boolean}      [vertical=true] - Identifies whether navigation shoud be vertical or horizontal
 * @property {string}       [idPrefix='id-'] - id prefix used for identifying individual node
 * @property {import('../constants').FieldMapping} fields - Field mapping to identify attributes to be used for state and identification of children
 */

/**
 * @typedef SwipeableOptions
 * @property {boolean} horizontal - Swipe horizontally
 * @property {boolean} vertical   - Swipe vertically
 * @property {boolean} enabled    - Enable swiping
 * @property {number}  threshold  - Threshold for swipe
 * @property {number}  minSpeed   - Minimum speed for swipe
 */

/**
 * @typedef TraversableOptions
 * @property {boolean} horizontal - Traverse horizontally
 * @property {boolean} nested     - Traverse nested items
 * @property {boolean} enabled    - Enable traversal
 * @property {string}  value      - Value to be used for traversal
 * @property {Array<*>} items     - An array containing the data set to traverse
 * @property {Array<integer} [indices] - Indices of the items to be traversed
 */

/**
 * @typedef PositionTracker
 * @property {integer} index
 * @property {integer} previousIndex
 */

/**
 * @typedef EventHandlers
 * @property {function} [keydown]
 * @property {function} [keyup]
 * @property {function} [click]
 * @property {function} [touchstart]
 * @property {function} [touchmove]
 * @property {function} [touchend]
 * @property {function} [touchcancel]
 * @property {function} [mousedown]
 * @property {function} [mouseup]
 * @property {function} [mousemove]
 */

/**
 * @typedef {Object} ActionHandlers
 * @property {Function} [next]
 * @property {Function} [previous]
 * @property {Function} [select]
 * @property {Function} [escape]
 * @property {Function} [collapse]
 * @property {Function} [expand]
 */

/**
 * @typedef {Object} NavigationOptions
 * @property {Boolean} [horizontal]
 * @property {Boolean} [nested]
 * @property {Boolean} [enabled]
 */

/**
 * @typedef {Object} KeyboardActions
 * @property {Function} [ArrowDown]
 * @property {Function} [ArrowUp]
 * @property {Function} [ArrowRight]
 * @property {Function} [ArrowLeft]
 * @property {Function} [Enter]
 * @property {Function} [Escape]
 * @property {Function} [" "]
 */

/**
 * @typedef {Object} TouchTracker
 * @property {number} startX - The start X position of the touch.
 * @property {number} startY - The start Y position of the touch.
 * @property {number} startTime - The start time of the touch.
 */

/**
 * @typedef {Object} PushDownOptions
 * @property {string} selector - The CSS selector for the child element to which keyboard events will be forwarded.
 * @property {Array<string>} [events=['keydown', 'keyup', 'keypress']] - The keyboard events to forward.
 */

/**
 * @typedef {Object} Bounds
 * @property {number} lower
 * @property {number} upper
 */
