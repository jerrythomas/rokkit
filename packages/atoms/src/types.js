/**
 * @typedef {'empty'|'last'|'child'|'sibling'} ConnectionType
 */
export default {}
