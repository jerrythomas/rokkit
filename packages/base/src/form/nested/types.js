import * as custom from '../custom'
import Input from '../Input.svelte'
import InputArray from './InputArray.svelte'
import InputObject from './InputObject.svelte'
import List from '../group/Select.svelte'

export const inputTypes = {
	input: Input,
	json: custom.InputJSON,
	array: InputArray,
	object: InputObject,
	radio: custom.RadioGroup,
	checkbox: custom.CheckBox,
	toggle: custom.Toggle,
	switch: custom.Switch,
	list: List
}
