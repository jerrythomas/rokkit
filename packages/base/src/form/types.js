import { default as Input } from './Input.svelte'
import * as base from './base'
import * as custom from './custom'
import * as nested from './nested'

export const wrappedInput = {
	text: base.InputText,
	number: base.InputNumber,
	email: base.InputEmail,
	tel: base.InputTel,
	color: base.InputColor,
	range: base.InputRange,
	date: base.InputDate,
	time: base.InputTime,
	datetime: base.InputDateTime,
	url: base.InputUrl,
	week: base.InputWeek,
	month: base.InputMonth,
	password: base.InputPassword,
	file: base.InputFile,
	textarea: base.TextArea,
	rating: custom.Rating
}
export const inputTypes = {
	input: Input,
	json: custom.InputJSON,
	array: nested.InputArray,
	object: nested.InputObject,
	radio: custom.RadioGroup,
	checkbox: custom.CheckBox,
	toggle: custom.Toggle,
	switch: custom.Switch
}
