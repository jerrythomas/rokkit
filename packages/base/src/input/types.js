import { default as InputText } from './InputText.svelte'
import { default as InputJSON } from './InputJSON.svelte'
import { default as InputArray } from './InputArray.svelte'
import { default as InputNumber } from './InputNumber.svelte'
import { default as InputObject } from './InputObject.svelte'
import { default as InputCheckBox } from './InputCheckBox.svelte'
import { default as InputRating } from './InputRating.svelte'
import { default as RadioGroup } from './RadioGroup.svelte'

export const inputTypes = {
	text: InputText,
	json: InputJSON,
	array: InputArray,
	number: InputNumber,
	object: InputObject,
	checkbox: InputCheckBox,
	rating: InputRating,
	radio: RadioGroup
}
