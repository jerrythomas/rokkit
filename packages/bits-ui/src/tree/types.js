/**
 * @typedef {'empty'|'last'|'child'|'sibling'} TreeLineType
 */

/**
 * @typedef {Object} NodeStateIcons
 * @property {import('svelte').ComponentType} [expanded] - Icon for expanded state
 * @property {import('svelte').ComponentType} [collapsed] - Icon for collapsed state
 * @property {import('svelte').ComponentType} [selected] - Icon for selected state
 * @property {import('svelte').ComponentType} [loading] - Icon for loading state
 */

/**
 * @typedef {Object} TreeContext
 * @property {import('@rokkit/states').NestedController} controller - Tree controller
 * @property {Array<any>} items - Tree items
 * @property {any} value - Selected value
 * @property {import('@rokkit/core').FieldMapping} fields - Field mappings
 * @property {NodeStateIcons} icons - Icons for different node states
 */

/**
 * @typedef {Object} TreeNodeEvent
 * @property {any} item - The node item
 * @property {any} value - The node value
 */

/**
 * @typedef {Object} TreeToggleEvent
 * @property {any} item - The node item
 * @property {boolean} expanded - Whether the node is expanded
 */

/**
 * @typedef {Object} TreeMoveEvent
 * @property {any} item - The node item
 * @property {string} direction - Move direction
 */
