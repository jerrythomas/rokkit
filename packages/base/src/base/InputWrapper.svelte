<script>
	import { inputs } from './index'

	export let value
	export let type = 'text'
</script>

<svelte:component this={inputs[type]} bind:value {...$$restProps} />
