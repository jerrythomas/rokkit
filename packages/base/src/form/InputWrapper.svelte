<script>
	import { wrappedInput } from './types'

	export let value
	export let type = 'text'
</script>

<svelte:component
	this={wrappedInput[type]}
	bind:value
	{...$$restProps}
	on:change
	on:focus
	on:blur
/>
