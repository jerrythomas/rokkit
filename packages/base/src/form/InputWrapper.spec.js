import { describe, expect, beforeEach, it } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import InputWrapper from './InputWrapper.svelte'
import { wrappedInput } from './types'

describe('InputWrapper.svelte', () => {
	const types = Object.keys(wrappedInput)
	const withProps = [
		['text', { minlength: 3, maxlength: 10, value: 'foo' }],
		['number', { min: 3, max: 10, value: 4 }],
		['rating', { max: 3, value: 2, name: 'rating' }]
	]
	beforeEach(() => cleanup())

	it('should render text input', () => {
		const { container } = render(InputWrapper, { value: null })
		const wrapper = container.childNodes[0]
		expect(wrapper.innerHTML).toEqual('<input type="text">')
	})

	it.each(types)('should render input for type "%s"', (type) => {
		const { container } = render(InputWrapper, { type, value: null })
		const wrapper = container.childNodes[0]
		expect(wrapper.childNodes[0]).toMatchSnapshot()
	})

	it.each(types.filter((type) => type !== 'rating'))(
		'should render input for type "%s" as readonly',
		(type) => {
			const { container } = render(InputWrapper, {
				type,
				value: null,
				readonly: true
			})
			const wrapper = container.childNodes[0]
			expect(wrapper.childNodes[0]).toMatchSnapshot()
		}
	)

	it.each(types)('should render input for type "%s" as required', (type) => {
		const { container } = render(InputWrapper, {
			type,
			value: null,
			readonly: true
		})
		const wrapper = container.childNodes[0]
		expect(wrapper.childNodes[0]).toMatchSnapshot()
	})

	it.each(withProps)(
		'should render input for type "%s" with custom props',
		(type, props) => {
			const { container } = render(InputWrapper, {
				type,
				...props
			})
			const wrapper = container.childNodes[0]
			expect(wrapper.childNodes[0]).toMatchSnapshot()
		}
	)

	// it('should render text input as required', () => {
	// 	const { container } = render(InputWrapper, {
	// 		value: 'foo',
	// 		required: true
	// 	})
	// 	const wrapperDiv = container.childNodes[0]
	// 	expect(wrapperDiv.innerHTML).toEqual('<input type="text" required="">')
	// })

	// it('should render text input with length limits', () => {
	// 	const { container } = render(InputWrapper, {
	// 		value: 'foo',
	// 		minlength: 3,
	// 		maxlength: 10
	// 	})
	// 	const wrapperDiv = container.childNodes[0]
	// 	expect(wrapperDiv.innerHTML).toEqual(
	// 		'<input type="text" minlength="3" maxlength="10">'
	// 	)
	// })
})
