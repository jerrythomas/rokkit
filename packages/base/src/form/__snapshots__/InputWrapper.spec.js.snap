// Vitest Snapshot v1

exports[`InputWrapper.svelte > should render input for type "color" 1`] = `
<input
  type="color"
/>
`;

exports[`InputWrapper.svelte > should render input for type "color" as readonly 1`] = `
<input
  readonly="true"
  type="color"
/>
`;

exports[`InputWrapper.svelte > should render input for type "color" as required 1`] = `
<input
  readonly="true"
  type="color"
/>
`;

exports[`InputWrapper.svelte > should render input for type "date" 1`] = `
<input
  type="date"
/>
`;

exports[`InputWrapper.svelte > should render input for type "date" as readonly 1`] = `
<input
  readonly="true"
  type="date"
/>
`;

exports[`InputWrapper.svelte > should render input for type "date" as required 1`] = `
<input
  readonly="true"
  type="date"
/>
`;

exports[`InputWrapper.svelte > should render input for type "datetime" 1`] = `
<input
  type="datetime-local"
/>
`;

exports[`InputWrapper.svelte > should render input for type "datetime" as readonly 1`] = `
<input
  readonly="true"
  type="datetime-local"
/>
`;

exports[`InputWrapper.svelte > should render input for type "datetime" as required 1`] = `
<input
  readonly="true"
  type="datetime-local"
/>
`;

exports[`InputWrapper.svelte > should render input for type "email" 1`] = `
<input
  type="number"
/>
`;

exports[`InputWrapper.svelte > should render input for type "email" as readonly 1`] = `
<input
  readonly="true"
  type="number"
/>
`;

exports[`InputWrapper.svelte > should render input for type "email" as required 1`] = `
<input
  readonly="true"
  type="number"
/>
`;

exports[`InputWrapper.svelte > should render input for type "file" 1`] = `
<input
  type="file"
/>
`;

exports[`InputWrapper.svelte > should render input for type "file" as readonly 1`] = `
<input
  readonly="true"
  type="file"
/>
`;

exports[`InputWrapper.svelte > should render input for type "file" as required 1`] = `
<input
  readonly="true"
  type="file"
/>
`;

exports[`InputWrapper.svelte > should render input for type "month" 1`] = `
<input
  type="month"
/>
`;

exports[`InputWrapper.svelte > should render input for type "month" as readonly 1`] = `
<input
  readonly="true"
  type="month"
/>
`;

exports[`InputWrapper.svelte > should render input for type "month" as required 1`] = `
<input
  readonly="true"
  type="month"
/>
`;

exports[`InputWrapper.svelte > should render input for type "number" 1`] = `
<input
  type="number"
/>
`;

exports[`InputWrapper.svelte > should render input for type "number" as readonly 1`] = `
<input
  readonly="true"
  type="number"
/>
`;

exports[`InputWrapper.svelte > should render input for type "number" as required 1`] = `
<input
  readonly="true"
  type="number"
/>
`;

exports[`InputWrapper.svelte > should render input for type "number" with custom props 1`] = `
<input
  max="10"
  min="3"
  type="number"
/>
`;

exports[`InputWrapper.svelte > should render input for type "password" 1`] = `
<input
  type="password"
/>
`;

exports[`InputWrapper.svelte > should render input for type "password" as readonly 1`] = `
<input
  readonly="true"
  type="password"
/>
`;

exports[`InputWrapper.svelte > should render input for type "password" as required 1`] = `
<input
  readonly="true"
  type="password"
/>
`;

exports[`InputWrapper.svelte > should render input for type "range" 1`] = `
<input
  type="range"
/>
`;

exports[`InputWrapper.svelte > should render input for type "range" as readonly 1`] = `
<input
  readonly="true"
  type="range"
/>
`;

exports[`InputWrapper.svelte > should render input for type "range" as required 1`] = `
<input
  readonly="true"
  type="range"
/>
`;

exports[`InputWrapper.svelte > should render input for type "rating" 1`] = `
<rating
  class="flex cursor-pointer select-none"
  tabindex="0"
>
   
  <icon
    class="rating-empty"
  />
  <icon
    class="rating-empty"
  />
  <icon
    class="rating-empty"
  />
  <icon
    class="rating-empty"
  />
  <icon
    class="rating-empty"
  />
</rating>
`;

exports[`InputWrapper.svelte > should render input for type "rating" as required 1`] = `
<rating
  class="flex cursor-pointer select-none"
  tabindex="0"
>
   
  <icon
    class="rating-empty"
  />
  <icon
    class="rating-empty"
  />
  <icon
    class="rating-empty"
  />
  <icon
    class="rating-empty"
  />
  <icon
    class="rating-empty"
  />
</rating>
`;

exports[`InputWrapper.svelte > should render input for type "rating" with custom props 1`] = `
<rating
  class="flex cursor-pointer select-none"
  tabindex="0"
>
  <input
    hidden=""
    max="3"
    min="0"
    name="rating"
    type="number"
  />
   
  <icon
    class="rating-filled"
  />
  <icon
    class="rating-filled"
  />
  <icon
    class="rating-empty"
  />
</rating>
`;

exports[`InputWrapper.svelte > should render input for type "tel" 1`] = `
<input
  type="url"
/>
`;

exports[`InputWrapper.svelte > should render input for type "tel" as readonly 1`] = `
<input
  readonly="true"
  type="url"
/>
`;

exports[`InputWrapper.svelte > should render input for type "tel" as required 1`] = `
<input
  readonly="true"
  type="url"
/>
`;

exports[`InputWrapper.svelte > should render input for type "text" 1`] = `
<input
  type="text"
/>
`;

exports[`InputWrapper.svelte > should render input for type "text" as readonly 1`] = `
<input
  readonly="true"
  type="text"
/>
`;

exports[`InputWrapper.svelte > should render input for type "text" as required 1`] = `
<input
  readonly="true"
  type="text"
/>
`;

exports[`InputWrapper.svelte > should render input for type "text" with custom props 1`] = `
<input
  maxlength="10"
  minlength="3"
  type="text"
/>
`;

exports[`InputWrapper.svelte > should render input for type "textarea" 1`] = `
<textarea
  type="week"
/>
`;

exports[`InputWrapper.svelte > should render input for type "textarea" as readonly 1`] = `
<textarea
  readonly="true"
  type="week"
/>
`;

exports[`InputWrapper.svelte > should render input for type "textarea" as required 1`] = `
<textarea
  readonly="true"
  type="week"
/>
`;

exports[`InputWrapper.svelte > should render input for type "time" 1`] = `
<input
  type="time"
/>
`;

exports[`InputWrapper.svelte > should render input for type "time" as readonly 1`] = `
<input
  readonly="true"
  type="time"
/>
`;

exports[`InputWrapper.svelte > should render input for type "time" as required 1`] = `
<input
  readonly="true"
  type="time"
/>
`;

exports[`InputWrapper.svelte > should render input for type "url" 1`] = `
<input
  type="url"
/>
`;

exports[`InputWrapper.svelte > should render input for type "url" as readonly 1`] = `
<input
  readonly="true"
  type="url"
/>
`;

exports[`InputWrapper.svelte > should render input for type "url" as required 1`] = `
<input
  readonly="true"
  type="url"
/>
`;

exports[`InputWrapper.svelte > should render input for type "week" 1`] = `
<input
  type="week"
/>
`;

exports[`InputWrapper.svelte > should render input for type "week" as readonly 1`] = `
<input
  readonly="true"
  type="week"
/>
`;

exports[`InputWrapper.svelte > should render input for type "week" as required 1`] = `
<input
  readonly="true"
  type="week"
/>
`;
